crates/proto/src/serialize/binary/encoder.rs

```diff
@@ -184,26 +192,65 @@ impl<'a> BinEncoder<'a> {
         self.mode
     }
 
-    /// If set to true, then names will be written into the buffer in canonical form
-    pub fn set_canonical_names(&mut self, canonical_names: bool) {
-        self.canonical_names = canonical_names;
+    /// If set to true, then records will be written into the buffer in DNSSEC canonical form
+    pub fn set_canonical_form(&mut self, canonical_form: bool) {
+        self.canonical_form = canonical_form;
+    }
+
+    /// Returns true if the encoder is writing in DNSSEC canonical form
+    pub fn is_canonical_form(&self) -> bool {
+        self.canonical_form
     }
 
-    /// Returns true if then encoder is writing in canonical form
-    pub fn is_canonical_names(&self) -> bool {
-        self.canonical_names
+    /// If set to true, then names may be encoded in compressed form
+    pub fn set_compression(&mut self, compression: bool) {
+        self.compression = compression;
     }
 
-    /// Emit all names in canonical form, useful for <https://tools.ietf.org/html/rfc3597>
-    pub fn with_canonical_names<F: FnOnce(&mut Self) -> ProtoResult<()>>(
+    /// Returns true if name compression is enabled
+    pub fn is_compression_enabled(&self) -> bool {
+        self.compression
+    }
+
+    /// If set to true, then names will be written into the buffer transformed to lowercase
+    pub fn set_lowercase_names(&mut self, lowercase_names: bool) {
+        self.lowercase_names = lowercase_names;
+    }
+
+    /// Returns true if the encoder is writing names transformed to lowercase
+    pub fn is_lowercase_names(&self) -> bool {
+        self.lowercase_names
+    }
```

Comment:
Are there really 8 useful modes to encode names in? I would expect that some of these collapse, and as such we should try to shape the API in a better way.

Reply:
There are actually three relevant modes for name encoding. The `canonical_form` flag does not directly affect name encoding, rather it is used when setting the two name encoding flags at the outset of encoding the RDATA. We need this flag in order to communicate from `Message::to_bytes()` or `TBS::new()` to RDATA types whether we're using canonical form or not. Considering just the compression and lowercase name flags, AFAIK we only use three out of four combinations, and nothing requires compressed lowercase names.